crates/telemetry/src/logs.rs

```diff
@@ -0,0 +1,116 @@
+use std::{ascii::escape_default, sync::OnceLock, time::Duration};
+
+use anyhow::bail;
+use opentelemetry::{
+    global,
+    logs::{Logger, LoggerProvider},
+};
+use opentelemetry_otlp::LogExporterBuilder;
+use opentelemetry_sdk::{
+    logs::{BatchConfigBuilder, BatchLogProcessor},
+    resource::{EnvResourceDetector, TelemetryResourceDetector},
+    Resource,
+};
+
+use crate::{
+    detector::SpinResourceDetector,
+    env::{self, otel_logs_enabled, OtlpProtocol},
+};
+
+/// Handle an application log. Has the potential to both forward the log to OTel and to emit it as a
+/// tracing event.
+pub fn handle_app_log(buf: &[u8]) {
+    app_log_to_otel(buf);
+    app_log_to_tracing_event(buf);
+}
+
+/// Forward the app log to OTel.
+fn app_log_to_otel(buf: &[u8]) {
+    if !otel_logs_enabled() {
+        return;
+    }
+
+    let logger = global::logger_provider().logger("spin");
+    if let Ok(s) = std::str::from_utf8(buf) {
+        logger.emit(
+            opentelemetry::logs::LogRecord::builder()
+                .with_body(s.to_owned())
+                .build(),
+        );
+    } else {
+        logger.emit(
+            opentelemetry::logs::LogRecord::builder()
+                .with_body(
+                    buf.iter()
+                        .take(50)
+                        .map(|&x| escape_default(x).to_string())
+                        .collect::<String>(),
+                )
+                .with_attribute("app_log_non_utf8", true)
+                .build(),
+        );
+    }
+}
+
+/// Takes a Spin application log and emits it as a tracing event. This acts as a compatibility layer
+/// to easily get Spin app logs as events in our OTel traces.
+fn app_log_to_tracing_event(buf: &[u8]) {
```

Comment:
Deja vu...didn't you just add this?

Reply:
It's appearing as new b/c I renamed `log.rs` to `logs.rs`

Reply:
It looks like you didn't commit the removal of `log.rs`, which will be part of why it's showing as new 😅

Reply:
Ohhhhh, I'll fix that.